### ISSUE TITLE
Add support for <--nextpage-->

### ISSUE THREAD

**Issue by shazahm1** [shazahm1/Easy-Table-of-Contents#13]

**Comment by shazahm1** [comment]
@unixtam 

Just a heads up... I've begun implementing this feature. I have taken a different direction in implementing the extraction of headings and TOC creation.

Right now this only affects the auto inserted TOC. The shortcode and widget still function, I think, I have not tested. If they do, they still are implemented using the "old" methods.

The `extract_headings()` and `build()` methods are very likely to be completely removed and/or not backwards compatible.

Wanted to let you know since I know you are using `extract_headings()` and perhaps other code which may no longer exist moving forward.

**Comment by unixtam** [comment]
@shazahm1, thanks for the heads-up! I'm sure the code will be modular enough that there's a way to build a toc for a given post somehow. Currently, my code looks like this:

```
function get_ez_toc_headings( $post ) {
	$find = $replace = [];
	$html = ezTOC::extract_headings( $find, $replace, $post );
	$html = str_replace( 'href="#', 'href="' . get_permalink( $post ) . '#', $html );
	return $html;
}
```

If you find the time, I'd appreciate an update once you've settled on a new code structure, so that I can update this accordingly.

**Comment by shazahm1** [comment]
@unixtam 

Ok, I think this is pretty close!

I think your function would need to become something like this:

```php
function get_ez_toc_headings( $post ) {
	$ezPost = new ezTOC_Post( $post );
	$ezPost->process(); // or $ezPost->applyContentFilter()->process(); if `the_content` filter needs to be applied to the post content
	$html = $ezPost->getTOCList();
	$html = str_replace( 'href="#', 'href="' . get_permalink( $post ) . '#', $html );
	return $html;
}
```

This function will break on paged posts as created with `<--nextpage-->`. This because absolute URLs are created for the pages that the user is **not** currently on within the paged post. Relative URLs are still created on the currently being viewed page of the post. So in this scenario, this function would add the permalink to the TOC anchor which already exists.

If you never create paged posts, this will not be an issue.

If I may, what is the purpose of making the in page anchors absolute instead of leaving them as relative. Perhap this is something that should be taken into account in the "core" code.

FYI, the following methods in ezTOC are now dead and will be removed:

 - `url_anchor_target()`
 - `build_hierarchy()`
 - `extract_headings()`
 - `build()`

These now, more or less, exist in the `ezTOC_Post` class. These have been refactored significantly enough that there is not really a 1-2-1 relationship between the old and new. Especially the old `extract_headings()` method compared to the new `extractHeadings()` in `ezTOC_Post`.

Any way, unless you find an issue... I think all that is left to do is remove the dead code.

**Comment by unixtam** [comment]
Thanks, @shazahm1, that almost worked out of the box (had to remove my `<ul>`).

> If you never create paged posts, this will not be an issue.

I probably won't. The only purpose I see for paged posts is getting a higher ads to content ratio, and since I'm not big on ads, I usually don't paginate my posts.

> If I may, what is the purpose of making the in page anchors absolute instead of leaving them as relative.

Sure, I'm aggregating the TOCs of several posts into a meta TOC on a separate page. It's a list of linked post titles, and below each title there's the linked post's TOC.

> Any way, unless you find an issue... I think all that is left to do is remove the dead code.

LGTM, remove away! :)

**Comment by shazahm1** [comment]
@unixtam 

Thanks for the feedback!

After I posted my last reply I had wondered if that is what you were doing...

Ideally if I could add logic so the TOC links were all absolute, that would work better for you? Perhaps add a setting.

**Comment by unixtam** [comment]
Full URLs for every named anchor would seem like a waste of bandwidth to me. Sure, you could make it an option, but is it worth the effort? I mean it would save me exactly one line of code that's already written. I'd just keep it as is. :)

**Comment by shazahm1** [comment]
Ok, this is ready for release. I'd recommend giving it one last quick test ;) because of the following:

 - Removed all dead code. 53bc192edb2be9f3b9840ab496575b9f6ff3209e
 - Removed some duplicated code by abstracting out into a couple private methods. f1d89649e4f87c69cd252d6a77506c91becb62d4
 - Add a small quick check on whether absolute or relative links should be created. d596b96fc832fcf0570720f6cfb81e1717d270eb
 - Added a few helper functions for devs. c5bbfc44dfadfd88d2f866dd77ac6351c5dc1a3e

In theory, your function is no longer needed. You should simple be able to do this:

```php
// Get TOC list.
$toc = get_ez_toc_list( $post ); // $post can be a WP_Post object or post ID.

// Display (echo) the TOC list.
ez_toc_list( $post ); // $post can be a WP_Post object or post ID.

// These helper functions are set to ensure `the_content` filter is applied. This can be disabled.
$toc = get_ez_toc_list( $post, false ); // or ez_toc_list( $post, false );
```

Because of d596b96fc832fcf0570720f6cfb81e1717d270eb you should not have to do the `str_replace()` on the output.

**Comment by shazahm1** [comment]
@unixtam 

Wondering if you have had a moment to give you specific needs a quick test with the latest changes? I'd like to push this out soon. But I don't want to break your usage.

**Comment by unixtam** [comment]
@shazahm1 thanks! I've been working on other things lately, so I hadn't tested your latest changes. I have now. On my meta TOC, the code from 2 weeks ago still works as expected, and so does your `ez_toc_list()` function, so good job on that!

However, I've noticed there's an issue with the TOCs on individual pages now, which do not include headings added via `the_content` filter (at priority 14). Before the update, these were included in the individual page's TOC and excluded on the meta TOC (which was perfect, since the extra headings are for notes and references). I took a glance at your code and saw that you are hooking `the_content` at priority 100, so that's not it, but something is definitely not working as expected. To test this, you could (probably) just add a `h2` heading via `the_content` filter at priority 14 and see whether it's included in the TOC.

**Comment by shazahm1** [comment]
Ok, on a preliminary check...

 - It seems the shortcode has not ever included headings added via the `the_content` filter. I checked the past three releases.
 - Presently, the auto insert does seem to include headings added via the `the_content` filter.
 - The widget had included headings added via the `the_content` filter. Now it does not, so this is a regression I'll address today.

I am not exactly sure what you mean by "*these were included in the individual page's TOC and excluded on the meta TOC*".

 - Which method are you using to insert the TOC on individual pages?
 - What do you mean by the "*meta TOC*"?
 - Are the headings added via `the_content` filter on the page where you are using your custom function/`ez_toc_list()`? They should be included by default.

**Comment by shazahm1** [comment]
Ok, I've corrected the widget regression in 18c644fa3c61f7790f47a78f9e0955c7fa8c34e3.

It seems a method static variable retains the last value across multiple instances of an object. I did not know that. I thought it only be retained per instance.

This prevented `the_content` filter from being applied in all instances after the first run on the method regardless on if it was a new instance or not.

I removed the static variable and the instanceof check because they are not really needed. Especially the instanceof check because the class requires an instance of `WP_Post` to be initialized.

Inserting the TOC via the shortcode still  does not include headings included via `the_content` filter. I consider that a bug. Running the filter within the shortcode callback seems to create an infinite loop. So I'm not sure these can be "captured". Perhaps removing the core WP filter which runs on `the_content` filter will work.

**Comment by shazahm1** [comment]
Changes made in 3b63a9afb384bb018326defbd11a2a60540c493a and b8614499bf436a14b1705f31c5d6550c9431dff4 ensure headings added to post content via `the_content` filter are also captured when using the shortcodes.

I think the only part that I am uncertain about now is how this will affect your meta TOC???

**Comment by unixtam** [comment]
Let's see.

> Which method are you using to insert the TOC on individual pages?

On the page I tested, that was auto-insert.

> What do you mean by the "meta TOC"?

The TOC of TOCs we have been talking about.

> Are the headings added via the_content filter on the page where you are using your custom function/ez_toc_list()? They should be included by default.

No, I'm calling `ez_toc_list()` in a loop to generate the "meta TOC" on a different page, in which case, I do not want the extra headings added, for which, I am guessing, I should be setting the second parameter of my `ez_toc_list()` call to `false`.

I have pulled your recent changes. Now the auto-inserted TOC on a test page correctly captures the custom headings as expected. The meta TOC page, however, is another story. The widget on that page produces a fatal error because maximum function nesting is being reached, meaning the `ez_toc_list()` call inside my "meta TOC" shortcode suddenly produces an infinite loop when the widget is active. When the widget is not active, the weirdness takes another form. Now the two extra headings from one of the invidivual posts are added to each of the subsequent TOCs and additionally added once more after the page's content. I can safely say, this is even more broken than before. ;)

Silver lining: When I do set the second parameter of `ez_toc_list()` call to `false` inside my "meta TOC" shortcode, the TOCs are rendering as expected. But, when I have the widget active on that page now, it produces two undefined index notices, both in `includes/class.post.php on line 726`.

**Comment by shazahm1** [comment]
Ok understand the meta TOC now.

I don't think it's more broken now... just need to figure out what's causing the infinite loop. I don't have enough info to duplicate your setup so I can replicate.

 - Could you give me the full code for your widget (or is this the TOC widget you're talking about???)
 - Could you give me the full code for your shortcode?
 - Copy of the text of the page content of the meta TOC.
 - Is this page public? Can you share a link?

Alternatively, which would be quicker, if I've earned enough trust, FTP access and WP admin access and I'm sure I could get this sorted very quickly. I do this almost every day for others in support of my primary plugin, Connections. It has commercial addons I sell and have to support.

I suspect that if you registered you own custom shortcode which is utilizing the ezTOC functions within the shortcode callback ... you need to remove it and then add it again before returning. Such as I did in 3b63a9afb384bb018326defbd11a2a60540c493a. OR set the second parameter of `ez_toc_list()` call to `false`. Perhaps do both. That will depends on if the expected headings are being captured or not. In you case, it sounds like they are. This would leave us with the two undefined index notices. But I am not certain if we're talking about the ezTOC widget or a widget you created.

**Comment by shazahm1** [comment]
Thanks, I was able to exactly replicate the issues you described.

The following code works:

```php
add_shortcode( 'meta_toc_test', function() {
    remove_shortcode( 'meta_toc_test' );
	$posts = get_posts( [
				'post_status' => 'publish',
				'numberposts' => 10,
				//'include'     => array( 7642, 9124, 9137 ),
				 //'post_type'   => 'page',
			] );
	
	ob_start();
?>
	<h1>this headline generates a notice if the widget is active</h1>
	<ul>
	<?php foreach ( $posts as $post ) : ?>
		<li>
			<a href="<?php the_permalink( $post ); ?>"><?php echo $post->post_title; ?></a>
			<?php ez_toc_list( $post ); ?>
		</li>
	<?php endforeach; ?>
	</ul>
<?php
	$output = ob_get_contents();
	ob_end_clean();
	return $output;
} );

add_shortcode( 'extra_headline_test', function() {
	global $extra_headline_test;
	$extra_headline_test = true;
} );

add_filter( 'the_content', function( $content ) {
	global $extra_headline_test;
	if ( !empty( $extra_headline_test ) ) {
		$content .= '<h2>Extra Headline</h2>';
	}
	return $content;
}, 14 );
```

The only line of code I added was `remove_shortcode( 'meta_toc_test' );`. I did change the posts query to grab my specific test posts. And, I did change `<?php ez_toc_list( $post, false ); ?>` to `<?php ez_toc_list( $post ); ?>`

If you were not using a closure, I would add the shortcode back at the end.

With these changes it operates without issue. Even if `ez_toc_list()` set to false.

That said, I'm going to see if I can resolve the index not found error that crops up if the `meta_toc_test` shortcode is not removed in the closure with `ez_toc_list()` set to false. Basically the code as you sent it.

**Comment by shazahm1** [comment]
I tracked down the  undefined index notices you are receiving... it was related to the number of required headings. If you go to the settings and change it to `1` the notices should no longer be present. I'll push a tweak in the next few moments which should correct this.

**Comment by shazahm1** [comment]
With the last change you should be able to use your code without issue!  :D

You only need to either add `remove_shortcode( 'meta_toc_test' );` to your closure or set the second parameter of `ez_toc_list()` call to `false`. Your choice.

**Comment by shazahm1** [comment]
Darn. There is still the duplicate heading issue.

**Comment by shazahm1** [comment]
Oh... the duplicate heading issue is more of a bug in your code.

It only takes the `extra_headline_test` to be run once to set a global var to `true`. After that, each subsequent time `the_content` filter is run, the extra headings will be added. In this case it'll be added to each time in the foreach loop where `ez_toc_list()` is called.

That said I would recommend adding `remove_shortcode( 'meta_toc_test' );` to your closure (or callback) and set the second parameter of `ez_toc_list()` call to `false`. 

I think that should button things up. Correct?

**Comment by unixtam** [comment]
Thanks for that! I've looked into it again. These are the results:

`remove_shortcode( 'meta_toc_test' );` does not seem to be a viable solution. When the eztoc widget is active, it removes the shortcode so that it's no longer available for the page content. Consequently, this only works when the widget is not used, in which case the infinite loop does not occur anyway. Maybe you could cache the headings for the current page, so that `the_content` filters run only once, regardless of whether or not the widget is active. That should prevent infinite loop issues and also save some execution time on pages where there is a TOC both in the page content and the widget.

https://github.com/shazahm1/Easy-Table-of-Contents/commit/a1dc1411a96f8387ae85ff1503e5135d53ff8b23 fixed the notices.

> Oh... the duplicate heading issue is more of a bug in your code.

You are right. That code was written under the assumption that `the_content` filters are only applied once per page load, which is not necessarily true. Thanks, I've updated the code to reset the global variables.

In summary: It works for me.

**Comment by shazahm1** [comment]
Thanks for the feedback! I'm planning on releasing this week.

In regards, even with cached processing, there's no way to guarantee that `the_content` filter will run once. Trust me.. in support my main plugin Connections Business Directory, I've run across countless instances where themes and other plugins will run `the_content` filter for various reasons. One example, Divi does this to pull the first image out of the post content.

You can "fix" your shortcode by using a function instead of of a closure, like this:

``` php
function meta_toc_test() {
    remove_shortcode( 'meta_toc_test' );
	$posts = get_posts( [
				'post_status' => 'publish',
				'numberposts' => 10,
				//'include'     => array( 7642, 9124, 9137 ),
				 //'post_type'   => 'page',
			] );
	
	ob_start();
?>
	<h1>this headline generates a notice if the widget is active</h1>
	<ul>
	<?php foreach ( $posts as $post ) : ?>
		<li>
			<a href="<?php the_permalink( $post ); ?>"><?php echo $post->post_title; ?></a>
			<?php ez_toc_list( $post ); ?>
		</li>
	<?php endforeach; ?>
	</ul>
<?php
	$output = ob_get_contents();
	ob_end_clean();
	add_shortcode( 'meta_toc_test', 'meta_toc_test' );
	return $output;
} );

add_shortcode( 'meta_toc_test', 'meta_toc_test' );
```

You register shortcode again right before you return the content like I had to in [`ezTOC_Post::applyContentFilter()`](https://github.com/shazahm1/Easy-Table-of-Contents/blob/master/includes/class.post.php#L112). I had to remove both the shortcodes and filters to ensure an infinite loop would not occur.

**Comment by okainov** [comment]
@shazahm1  why is this closed? It seems that there is still no support for paginated posts (at least in official released version).

What I see currently that TOC is here, but all links are only for the current page and do not respect other pages.

**Comment by okainov** [comment]
Small update:
Latest released version is 1.7.

Just checked with `master` - I can see some changes, but in general the whole TOC is broken, it misses almost half of my headings. However existing links now include page numbers and at least work. But I reverted it back to 1.7.

Pls reopen. Thanks!

**Comment by shazahm1** [comment]
@okainov 

That is because I have not had the opportunity to release this version publically. Feel free to pull the master branch from here or grab the RC version from here:

https://downloads.wordpress.org/plugin/easy-table-of-contents.2.0-rc1.zip

**Comment by okainov** [comment]
@shazahm1 see my updated commend

> Just checked with master - I can see some changes, but in general the whole TOC is broken, it misses almost half of my headings. However existing links now include page numbers and at least work. But I reverted it back to 1.7.

**Comment by shazahm1** [comment]
>  I can see some changes, but in general the whole TOC is broken, it misses almost half of my headings.

Please provide more specific details such as:

 - A link to the page with the issue.
 - A screenshot showing both the global settings and the post settings for ezTOC.

> However existing links now include page numbers and at least work. 

Not sure what you mean by this. ezTOC does not display page numbers.

**Comment by okainov** [comment]
@shazahm1 

With 1.7 Im able to thee the whole TOC.
I use toc-widget.

Page: https://olegu.ru/reports/2016/norway/norvegiya-v-ijule-solo-na-avto-s-palatkoj/

> Not sure what you mean by this. ezTOC does not display page numbers.

Here is what I mean: 
![image](https://user-images.githubusercontent.com/918446/43465040-bf8bf946-94dc-11e8-8a7f-d7fcb4a1df2e.png)



Example: ![image](https://user-images.githubusercontent.com/918446/43464673-f9bfe150-94db-11e8-8b47-4cfbc118b927.png)

My settings are in Russian, sorry:
General: ![image](https://user-images.githubusercontent.com/918446/43464725-1958b6ea-94dc-11e8-82e4-bc28a409455a.png)
Advanced: ![image](https://user-images.githubusercontent.com/918446/43464789-3434ebf0-94dc-11e8-8cec-0e644866bd5b.png)

**Comment by shazahm1** [comment]
@okainov 

I've opened two new issues and have asked additional questions, can you please reply back. Here are the links:

- #53
- #54

**Comment by olivier-duval** [comment]
Hello
I found how to manage the . The problem lies in the fact that the minimum number of headings is counted "by page" and not for the whole article. So the fix is to edit the class.post.php and
- add a member: private $TOCsize = 0;
- edit the private function extractHeadings( $content ) (lines 404) to only count the headings and not filter
- edit the private function processPages() (lines 259) to filter when the minimum is not met

See sources attached

However I still have problems with the hierarchical ul/li in multi-page  :-(

[class.post.php.txt](https://github.com/shazahm1/Easy-Table-of-Contents/files/6893606/class.post.php.txt)
